### PR TITLE
[Libomp] Place generated OpenMP headers into build resource directory

### PR DESCRIPTION
**Pull request by jhuber6**
Summary:
These headers are a part of the compiler's resource directory once
installed. However, they are currently placed in the binary directory
temporarily. This makes it more difficult to use the compiler out of the
build directory and will cause issues when moving to `liboffload`. This
patch changes the logic to write these instead to the copmiler's
resource directory inside of the build tree.

NOTE: This doesn't change the Fortran headers, I don't know enough about
those and it won't use the same directory.

### CODE REVIEW

File: openmp/runtime/src/CMakeLists.txt [hunk 1]
```diff
@@ -419,15 +426,15 @@ else()
...
   DESTINATION ${LIBOMP_HEADERS_INSTALL_PATH}
 )
 if(${LIBOMP_OMPT_SUPPORT})
-  install(FILES ${CMAKE_CURRENT_BINARY_DIR}/omp-tools.h DESTINATION ${LIBOMP_HEADERS_INSTALL_PATH})
+  install(FILES ${LIBOMP_HEADERS_INTDIR}/omp-tools.h DESTINATION ${LIBOMP_HEADERS_INSTALL_PATH})
   # install under legacy name ompt.h
-  install(FILES ${CMAKE_CURRENT_BINARY_DIR}/omp-tools.h DESTINATION ${LIBOMP_HEADERS_INSTALL_PATH} RENAME ompt.h)
-  set(LIBOMP_OMP_TOOLS_INCLUDE_DIR ${CMAKE_CURRENT_BINARY_DIR} PARENT_SCOPE)
+  install(FILES ${LIBOMP_HEADERS_INTDIR}/omp-tools.h DESTINATION ${LIBOMP_HEADERS_INSTALL_PATH} RENAME ompt.h)
+  set(LIBOMP_OMP_TOOLS_INCLUDE_DIR ${LIBOMP_HEADERS_INTDIR} PARENT_SCOPE)
```
Comment (jhuber6) [inline openmp/runtime/src/CMakeLists.txt:437]: FYI this is only required if the build is standalone now. The regular build will always have these included via the resource dir now.

**Comment by mjklemm** [comment]
If this PR can wait a bit, I can have a look at the Fortran part for this.

**Comment by mjklemm** [comment]
Ok. Then let's move the Fortran part to a different PR.

**Comment by shiltian** [comment]
IIRC there is a case where `omp.h` as well as other headers are installed into clang's directory.

File: openmp/runtime/src/CMakeLists.txt [hunk 2]
```diff
@@ -10,12 +10,19 @@
 
 include(ExtendPath)
 
+# The generated headers will be placed in clang's resource directory if present.
+if(${OPENMP_STANDALONE_BUILD})
+  set(LIBOMP_HEADERS_INTDIR ${CMAKE_CURRENT_BINARY_DIR})
+else()
+  set(LIBOMP_HEADERS_INTDIR ${LLVM_BINARY_DIR}/${LIBOMP_HEADERS_INSTALL_PATH})
```
Comment (shiltian) [inline openmp/runtime/src/CMakeLists.txt:17]: Is `LLVM_BINARY_DIR` still valid if user only does `-DLLVM_ENABLE_PROJECTS="openmp"`? In this case LLVM will not be probably built.

**Comment by jhuber6** [comment]
> IIRC there is a case where `omp.h` as well as other headers are installed into clang's directory.

These are currently installed into clang's directory, this patch just puts it in the resource directory for the build.

File: clang/test/Headers/Inputs/include/stdint.h [hunk 3]
```diff
@@ -16,4 +16,12 @@ typedef unsigned __INTPTR_TYPE__ uintptr_t;
...
 #endif
 
+#ifdef __INTPTR_MAX__
+#define  INTPTR_MAX    __INTPTR_MAX__
+#endif
+
+#ifdef __UINTPTR_MAX__
+#define UINTPTR_MAX   __UINTPTR_MAX__
+#endif
+
```
Comment (jplehr) [inline clang/test/Headers/Inputs/include/stdint.h:26]: Are these changes required to make this work?

**Review by jplehr — COMMENTED** [review]
The changes seem reasonable to me.

File: clang/test/Headers/Inputs/include/stdint.h (hunk 3 above)
Comment (jhuber6) [reply clang/test/Headers/Inputs/include/stdint.h:26]: Alternatively we could just delete this `stdint.h` altogether, because this only happens because it's using this one instead. But I assume that we override the regular `stdint.h` to reduce the number of live definitions in tests or something.

File: openmp/runtime/src/CMakeLists.txt (hunk 2 above)
Comment (shiltian) [reply openmp/runtime/src/CMakeLists.txt:17]: Is it `INSTDIR` or `INTDIR`? What does `INTDIR` mean?

Comment (jhuber6) [reply openmp/runtime/src/CMakeLists.txt:17]: Done to match `LLVM_RUNTIME_OUTPUT_INTDIR` conceptually, It's just intermediate directory. I.e. not where it's being installed, but where it's stored in the build.

Comment (shiltian) [reply openmp/runtime/src/CMakeLists.txt:17]: oh "intermediate"

File: clang/test/Headers/Inputs/include/stdint.h (hunk 3 above)
Comment (shiltian) [reply clang/test/Headers/Inputs/include/stdint.h:26]: This header is useful when invoking front end directly such that we don't need to pass things like `-internal-isystem` which usually are added by compiler driver.

**Review by shiltian — APPROVED** [review]
LGTM

**Comment by psteinfeld** [comment]
Hey, @jhuber.  This change broke the flang build.  

When I revert the change, everything builds cleanly.  So I reverted it in merge request 

After this change was added, I started getting messages like:
```
[4/205] Building CXX object projects/openmp/libomptarget/plugins-nextgen/host/CMakeFiles/omptarget.rtl.x86_64.dir/dynamic_ffi/ffi.cpp.o
FAILED: projects/openmp/libomptarget/plugins-nextgen/host/CMakeFiles/omptarget.rtl.x86_64.dir/dynamic_ffi/ffi.cpp.o 
/usr/bin/c++ -DDEBUG_PREFIX="\"TARGET x86_64 RTL\"" -DGTEST_HAS_RTTI=0 -DLIBOMPTARGET_NEXTGEN_GENERIC_PLUGIN_TRIPLE=\"x86_64-pc-linux-gnu\" -DOMPT_SUPPORT=1 -DTARGET_ELF_ID=EM_X86_64 -DTARGET_NAME=x86_64 -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/local/home/psteinfeld/up/build/projects/openmp/libomptarget/plugins-nextgen/host -I/local/home/psteinfeld/up/llvm-project/openmp/libomptarget/plugins-nextgen/host -I/local/home/psteinfeld/up/build/include -I/local/home/psteinfeld/up/llvm-project/llvm/include -I/local/home/psteinfeld/up/build/lib/clang/19/include -I/local/home/psteinfeld/up/llvm-project/openmp/libomptarget/plugins-nextgen/host/dynamic_ffi -I/local/home/psteinfeld/up/llvm-project/openmp/libomptarget/include -I/local/home/psteinfeld/up/llvm-project/openmp/libomptarget/plugins-nextgen/common/include -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -fno-lifetime-dse -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-uninitialized -Wno-nonnull -Wno-class-memaccess -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wno-comment -Wno-misleading-indentation -fdiagnostics-color -ffunction-sections -fdata-sections -Wall -Wcast-qual -Wimplicit-fallthrough -Wsign-compare -Wno-extra -Wno-pedantic -Wno-maybe-uninitialized -fno-semantic-interposition -fdata-sections -O3 -DNDEBUG -std=c++17 -fPIC -fvisibility=protected  -fno-exceptions -funwind-tables -fno-rtti -fno-exceptions -funwind-tables -fno-rtti -UNDEBUG -fno-exceptions -fno-rtti -MD -MT projects/openmp/libomptarget/plugins-nextgen/host/CMakeFiles/omptarget.rtl.x86_64.dir/dynamic_ffi/ffi.cpp.o -MF projects/openmp/libomptarget/plugins-nextgen/host/CMakeFiles/omptarget.rtl.x86_64.dir/dynamic_ffi/ffi.cpp.o.d -o projects/openmp/libomptarget/plugins-nextgen/host/CMakeFiles/omptarget.rtl.x86_64.dir/dynamic_ffi/ffi.cpp.o -c /local/home/psteinfeld/up/llvm-project/openmp/libomptarget/plugins-nextgen/host/dynamic_ffi/ffi.cpp
In file included from /usr/include/wchar.h:35,
                 from /usr/include/c++/8/cwchar:44,
                 from /usr/include/c++/8/bits/postypes.h:40,
                 from /usr/include/c++/8/bits/char_traits.h:40,
                 from /usr/include/c++/8/string:40,
                 from /local/home/psteinfeld/up/llvm-project/llvm/include/llvm/Support/DynamicLibrary.h:16,
                 from /local/home/psteinfeld/up/llvm-project/openmp/libomptarget/plugins-nextgen/host/dynamic_ffi/ffi.cpp:13:
/local/home/psteinfeld/up/build/lib/clang/19/include/stddef.h:31:42: error: missing binary operator before token "("
 #if !defined(__STDDEF_H) || __has_feature(modules) ||                          \
                                          ^
In file included from /usr/include/wchar.h:38,
                 from /usr/include/c++/8/cwchar:44,
                 from /usr/include/c++/8/bits/postypes.h:40,
                 from /usr/include/c++/8/bits/char_traits.h:40,
                 from /usr/include/c++/8/string:40,
                 from /local/home/psteinfeld/up/llvm-project/llvm/include/llvm/Support/DynamicLibrary.h:16,
                 from /local/home/psteinfeld/up/llvm-project/openmp/libomptarget/plugins-nextgen/host/dynamic_ffi/ffi.cpp:13:
/local/home/psteinfeld/up/build/lib/clang/19/include/stdarg.h:31:42: error: missing binary operator before token "("
```

**Comment by jhuber6** [comment]
> Hey, @jhuber. This change broke the flang build.
> 
> When I revert the change, everything builds cleanly. So I reverted it in merge request #88083.
> 
> After this change was added, I started getting messages like:
> 
> ```
> [4/205] Building CXX object projects/openmp/libomptarget/plugins-nextgen/host/CMakeFiles/omptarget.rtl.x86_64.dir/dynamic_ffi/ffi.cpp.o
> FAILED: projects/openmp/libomptarget/plugins-nextgen/host/CMakeFiles/omptarget.rtl.x86_64.dir/dynamic_ffi/ffi.cpp.o 
> /usr/bin/c++ -DDEBUG_PREFIX="\"TARGET x86_64 RTL\"" -DGTEST_HAS_RTTI=0 -DLIBOMPTARGET_NEXTGEN_GENERIC_PLUGIN_TRIPLE=\"x86_64-pc-linux-gnu\" -DOMPT_SUPPORT=1 -DTARGET_ELF_ID=EM_X86_64 -DTARGET_NAME=x86_64 -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/local/home/psteinfeld/up/build/projects/openmp/libomptarget/plugins-nextgen/host -I/local/home/psteinfeld/up/llvm-project/openmp/libomptarget/plugins-nextgen/host -I/local/home/psteinfeld/up/build/include -I/local/home/psteinfeld/up/llvm-project/llvm/include -I/local/home/psteinfeld/up/build/lib/clang/19/include -I/local/home/psteinfeld/up/llvm-project/openmp/libomptarget/plugins-nextgen/host/dynamic_ffi -I/local/home/psteinfeld/up/llvm-project/openmp/libomptarget/include -I/local/home/psteinfeld/up/llvm-project/openmp/libomptarget/plugins-nextgen/common/include -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -fno-lifetime-dse -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-uninitialized -Wno-nonnull -Wno-class-memaccess -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wno-comment -Wno-misleading-indentation -fdiagnostics-color -ffunction-sections -fdata-sections -Wall -Wcast-qual -Wimplicit-fallthrough -Wsign-compare -Wno-extra -Wno-pedantic -Wno-maybe-uninitialized -fno-semantic-interposition -fdata-sections -O3 -DNDEBUG -std=c++17 -fPIC -fvisibility=protected  -fno-exceptions -funwind-tables -fno-rtti -fno-exceptions -funwind-tables -fno-rtti -UNDEBUG -fno-exceptions -fno-rtti -MD -MT projects/openmp/libomptarget/plugins-nextgen/host/CMakeFiles/omptarget.rtl.x86_64.dir/dynamic_ffi/ffi.cpp.o -MF projects/openmp/libomptarget/plugins-nextgen/host/CMakeFiles/omptarget.rtl.x86_64.dir/dynamic_ffi/ffi.cpp.o.d -o projects/openmp/libomptarget/plugins-nextgen/host/CMakeFiles/omptarget.rtl.x86_64.dir/dynamic_ffi/ffi.cpp.o -c /local/home/psteinfeld/up/llvm-project/openmp/libomptarget/plugins-nextgen/host/dynamic_ffi/ffi.cpp
> In file included from /usr/include/wchar.h:35,
>                  from /usr/include/c++/8/cwchar:44,
>                  from /usr/include/c++/8/bits/postypes.h:40,
>                  from /usr/include/c++/8/bits/char_traits.h:40,
>                  from /usr/include/c++/8/string:40,
>                  from /local/home/psteinfeld/up/llvm-project/llvm/include/llvm/Support/DynamicLibrary.h:16,
>                  from /local/home/psteinfeld/up/llvm-project/openmp/libomptarget/plugins-nextgen/host/dynamic_ffi/ffi.cpp:13:
> /local/home/psteinfeld/up/build/lib/clang/19/include/stddef.h:31:42: error: missing binary operator before token "("
>  #if !defined(__STDDEF_H) || __has_feature(modules) ||                          \
>                                           ^
> In file included from /usr/include/wchar.h:38,
>                  from /usr/include/c++/8/cwchar:44,
>                  from /usr/include/c++/8/bits/postypes.h:40,
>                  from /usr/include/c++/8/bits/char_traits.h:40,
>                  from /usr/include/c++/8/string:40,
>                  from /local/home/psteinfeld/up/llvm-project/llvm/include/llvm/Support/DynamicLibrary.h:16,
>                  from /local/home/psteinfeld/up/llvm-project/openmp/libomptarget/plugins-nextgen/host/dynamic_ffi/ffi.cpp:13:
> /local/home/psteinfeld/up/build/lib/clang/19/include/stdarg.h:31:42: error: missing binary operator before token "("
> ```

So the error is caused by your `c++` not supporting `__has_feature` presumably. However I have no clue why that would happen due to this patch. I believe in a few days @jdoerfert will flip the switch and we won't even allow `libomptarget` to be built without using the runtimes support so this might not be worth thinking about too hard.

Any suggestions for reproducing it?

**Comment by jhuber6** [comment]
> Hey, @jhuber. This change broke the flang build.
> 
> When I revert the change, everything builds cleanly. So I reverted it in merge request #88083.
> 
> After this change was added, I started getting messages like:
> 
> ```
> [4/205] Building CXX object projects/openmp/libomptarget/plugins-nextgen/host/CMakeFiles/omptarget.rtl.x86_64.dir/dynamic_ffi/ffi.cpp.o
> FAILED: projects/openmp/libomptarget/plugins-nextgen/host/CMakeFiles/omptarget.rtl.x86_64.dir/dynamic_ffi/ffi.cpp.o 
> /usr/bin/c++ -DDEBUG_PREFIX="\"TARGET x86_64 RTL\"" -DGTEST_HAS_RTTI=0 -DLIBOMPTARGET_NEXTGEN_GENERIC_PLUGIN_TRIPLE=\"x86_64-pc-linux-gnu\" -DOMPT_SUPPORT=1 -DTARGET_ELF_ID=EM_X86_64 -DTARGET_NAME=x86_64 -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/local/home/psteinfeld/up/build/projects/openmp/libomptarget/plugins-nextgen/host -I/local/home/psteinfeld/up/llvm-project/openmp/libomptarget/plugins-nextgen/host -I/local/home/psteinfeld/up/build/include -I/local/home/psteinfeld/up/llvm-project/llvm/include -I/local/home/psteinfeld/up/build/lib/clang/19/include -I/local/home/psteinfeld/up/llvm-project/openmp/libomptarget/plugins-nextgen/host/dynamic_ffi -I/local/home/psteinfeld/up/llvm-project/openmp/libomptarget/include -I/local/home/psteinfeld/up/llvm-project/openmp/libomptarget/plugins-nextgen/common/include -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -fno-lifetime-dse -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-uninitialized -Wno-nonnull -Wno-class-memaccess -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wno-comment -Wno-misleading-indentation -fdiagnostics-color -ffunction-sections -fdata-sections -Wall -Wcast-qual -Wimplicit-fallthrough -Wsign-compare -Wno-extra -Wno-pedantic -Wno-maybe-uninitialized -fno-semantic-interposition -fdata-sections -O3 -DNDEBUG -std=c++17 -fPIC -fvisibility=protected  -fno-exceptions -funwind-tables -fno-rtti -fno-exceptions -funwind-tables -fno-rtti -UNDEBUG -fno-exceptions -fno-rtti -MD -MT projects/openmp/libomptarget/plugins-nextgen/host/CMakeFiles/omptarget.rtl.x86_64.dir/dynamic_ffi/ffi.cpp.o -MF projects/openmp/libomptarget/plugins-nextgen/host/CMakeFiles/omptarget.rtl.x86_64.dir/dynamic_ffi/ffi.cpp.o.d -o projects/openmp/libomptarget/plugins-nextgen/host/CMakeFiles/omptarget.rtl.x86_64.dir/dynamic_ffi/ffi.cpp.o -c /local/home/psteinfeld/up/llvm-project/openmp/libomptarget/plugins-nextgen/host/dynamic_ffi/ffi.cpp
> In file included from /usr/include/wchar.h:35,
>                  from /usr/include/c++/8/cwchar:44,
>                  from /usr/include/c++/8/bits/postypes.h:40,
>                  from /usr/include/c++/8/bits/char_traits.h:40,
>                  from /usr/include/c++/8/string:40,
>                  from /local/home/psteinfeld/up/llvm-project/llvm/include/llvm/Support/DynamicLibrary.h:16,
>                  from /local/home/psteinfeld/up/llvm-project/openmp/libomptarget/plugins-nextgen/host/dynamic_ffi/ffi.cpp:13:
> /local/home/psteinfeld/up/build/lib/clang/19/include/stddef.h:31:42: error: missing binary operator before token "("
>  #if !defined(__STDDEF_H) || __has_feature(modules) ||                          \
>                                           ^
> In file included from /usr/include/wchar.h:38,
>                  from /usr/include/c++/8/cwchar:44,
>                  from /usr/include/c++/8/bits/postypes.h:40,
>                  from /usr/include/c++/8/bits/char_traits.h:40,
>                  from /usr/include/c++/8/string:40,
>                  from /local/home/psteinfeld/up/llvm-project/llvm/include/llvm/Support/DynamicLibrary.h:16,
>                  from /local/home/psteinfeld/up/llvm-project/openmp/libomptarget/plugins-nextgen/host/dynamic_ffi/ffi.cpp:13:
> /local/home/psteinfeld/up/build/lib/clang/19/include/stdarg.h:31:42: error: missing binary operator before token "("
> ```

In any case, the `flang` bot is going to need to be changed very soon if this is an error so I would recommend pinging whoever's in charge of that. The recommended method is `LLVM_ENABLE_RUNTIMES=openmp`.

**Comment by jhuber6** [comment]
> So the error is caused by your `c++` not supporting `__has_feature` presumably. However I have no clue why that would happen due to this patch. I believe in a few days @jdoerfert will flip the switch and we won't even allow `libomptarget` to be built without using the runtimes support so this might not be worth thinking about too hard.
> 
> Any suggestions for reproducing it?

Oh, I think I get it. We're adding the resource directory include to the include paths. This is not conflicting with the ones from GCC. Honestly we only need this for `ompt` stuff which I could just do directly I suppose.

**Comment by psteinfeld** [comment]
I should have mentioned, I'm building with GCC 9.3.0.  Most of my team uses this compiler.

**Comment by jhuber6** [comment]
> I should have mentioned, I'm building with GCC 9.3.0. Most of my team uses this compiler.

Yeah, that's fine for building LLVM but in the future we'll probably require building the offloading runtime with an up-to-date clang. I think for now I can just restrict this patch to runtimes builds.
